# Python commands

We study some python commands that are important for doing approximations and Taylor series.

#### __Variable Assignment Statements__ 
We use the equal sign (=) to assign a value to a variable. Example to assign the value 10 to $x$ and the value 8 to Nguyen we do the following

In [1]:
x = 10
Nguyen = 8

The print command is used to print the output (i.e. display the result) of a program. 
Note: print should be written in lower case letters.  See below for examples. 

In [2]:
print Nguyen


In [3]:
print x

Just typing "Nguyen" would also display it's value. The question is why use print? As you will see later in this lecture, the print command allows you to display your result in any format you want.

In [4]:
Nguyen

8

#### __Rules for Variable Names__
1. Variables names must start with a letter or an underscore

2. The remainder of your variable name may consist of letters, numbers and underscores.

3. Variable names are case sentitive
_Example: NGUYEN, NGuyen, nguyen are each a different variable._

4. Descriptive names are very useful. If you are writing a program that sums the number of apples and grapes, then use apples and grapes as variable names.

5. Example of valid variable names: 
_val, val_2, _number, NUMBERS, Kristen, Fabio etc._

You can't use Python keywords as variable names. Example _print_ is a Python keyword, thus you can't use _print_ as a variable name. You get an error message if you try to do this. The error will state that the syntax is invalid. 

There is a Python module called keyword. It has a function called kwlist (i.e. a list of keywords). Importing keyword and calling kwlist will return a list of Python’s keywords. See below for an example.

In [5]:
import keyword
>>> keyword.kwlist

['and',
 'as',
 'assert',
 'break',
 'class',
 'continue',
 'def',
 'del',
 'elif',
 'else',
 'except',
 'exec',
 'finally',
 'for',
 'from',
 'global',
 'if',
 'import',
 'in',
 'is',
 'lambda',
 'not',
 'or',
 'pass',
 'print',
 'raise',
 'return',
 'try',
 'while',
 'with',
 'yield']

You should learn the Python keywords and avoid using them as variable names.

#### Basic operations
"+" is used for addition, "-" for subtraction, "*" for multiplication, "/" for division, "%" for remainder and "**" for exponent.

If both inputs are integers, the division operator "/" works as an integer division. Example 8/5 returns 1. If you desire an answer that is not a whole number, you must input a floating point number with decimal points. Example 8./5 or 8/5.  returns the value 1.6. This is demonstrated below.


In [6]:
val_div =8/5
print val_div

1


In [7]:
val_div =8./5
print val_div

1.6


Given that val_div has been assigned twice. What is the current value of val_div? Note that it takes the recent value assigned during the last execution of the program. It may not necessarily follow the order it appears in the program. 

In [8]:
print val_div

1.6


To increment the value of a variable, type the variable followed by "+=" and then the amount you wish to add.  Example x += 2 assigns 12 to x, since "x" was previously assigned the value 10. 

In [9]:
x += 2

In [10]:
print x

12


Note that every time you run the cell with x += 2 it increments the value of x by 2. Be sure to restart the current kernel and re-execute the whole notebook if you get results other than what you were expecting.

In the last lecture, we imported and used the "matplotlib.pyplot" and "numpy" module. Another module is "math". The dir() command shows the built-in functions available in a module. First, import the module and see what functions are available in it using dir(). An example is shown below.

In [11]:
import math

In [12]:
dir(math)

['__doc__',
 '__file__',
 '__name__',
 '__package__',
 'acos',
 'acosh',
 'asin',
 'asinh',
 'atan',
 'atan2',
 'atanh',
 'ceil',
 'copysign',
 'cos',
 'cosh',
 'degrees',
 'e',
 'erf',
 'erfc',
 'exp',
 'expm1',
 'fabs',
 'factorial',
 'floor',
 'fmod',
 'frexp',
 'fsum',
 'gamma',
 'hypot',
 'isinf',
 'isnan',
 'ldexp',
 'lgamma',
 'log',
 'log10',
 'log1p',
 'modf',
 'pi',
 'pow',
 'radians',
 'sin',
 'sinh',
 'sqrt',
 'tan',
 'tanh',
 'trunc']

In [13]:
import numpy as np

In [14]:
dir(np)

['ALLOW_THREADS',
 'AxisError',
 'BUFSIZE',
 'CLIP',
 'ComplexWarning',
 'DataSource',
 'ERR_CALL',
 'ERR_DEFAULT',
 'ERR_IGNORE',
 'ERR_LOG',
 'ERR_PRINT',
 'ERR_RAISE',
 'ERR_WARN',
 'FLOATING_POINT_SUPPORT',
 'FPE_DIVIDEBYZERO',
 'FPE_INVALID',
 'FPE_OVERFLOW',
 'FPE_UNDERFLOW',
 'False_',
 'Inf',
 'Infinity',
 'MAXDIMS',
 'MAY_SHARE_BOUNDS',
 'MAY_SHARE_EXACT',
 'MachAr',
 'ModuleDeprecationWarning',
 'NAN',
 'NINF',
 'NZERO',
 'NaN',
 'PINF',
 'PZERO',
 'PackageLoader',
 'RAISE',
 'RankWarning',
 'SHIFT_DIVIDEBYZERO',
 'SHIFT_INVALID',
 'SHIFT_OVERFLOW',
 'SHIFT_UNDERFLOW',
 'ScalarType',
 'Tester',
 'TooHardError',
 'True_',
 'UFUNC_BUFSIZE_DEFAULT',
 'UFUNC_PYVALS_NAME',
 'VisibleDeprecationWarning',
 'WRAP',
 '_NoValue',
 '__NUMPY_SETUP__',
 '__all__',
 '__builtins__',
 '__config__',
 '__doc__',
 '__file__',
 '__git_revision__',
 '__mkl_version__',
 '__name__',
 '__package__',
 '__path__',
 '__version__',
 '_distributor_init',
 '_globals',
 '_import_tools',
 '_mat',
 '_mklinit'

#### Functions

Here we will learn how to write a Python function. Python functions are defined using the "def" keyword. The "def" command is followed by the function name and parentheses. 

Any input parameters or arguments should be placed within these parentheses. 

The code block within every function starts with a colon (:) and is indented.

To indicate a block of code in Python, you must indent each line of the block by the same amount.

An example is shown below.

In [15]:
def squared(t):
    num_squared = t**2
    return num_squared

The function in the cell above is call "squared". It takes a number called "t" and returns its square. Here is an example of a code that uses "squared" to compute the square of the number 10. 

The statement "return num_squared" exits the function, and passes back the value of num_squared to the caller. A return statement with no arguments is the same as return None.

In [16]:
squared(10)

100

Here is another example showing a Python function that uses the if-elif-else statement. 

After Python gets age from the user, it enters the if-elif-else statement and checks each condition one after the other in the order they are given. So first it checks if age is less than 2, and if so, it indicates that the ticket is free and jumps out of the elif-else condition. If age is not less than 2, then it checks the next elif-condition to see if age is between 2 and 13. If so, it prints the appropriate message and jumps out of the if/elif-else statement. If neither the if-condition nor the elif-condition is True, then it executes the code in the else-block.

In [17]:
def checking(age):
    if age <= 2: 
        return 'your ticket is free'
    elif 2 < age < 13:
        return 'Ticket price: child fare is $20 '
    else:
        return 'Ticket price: adult fare is $50'

The following code executes the function checking() with the input argument age=45.

In [18]:
checking(45)

'Ticket price: adult fare is $50'

#### Additional examples: "print" command:



In [19]:
myval = 1.23456789
print "myval=%1.5f" %myval

myval=1.23457


In [20]:
myval = 1.23456789
print "myval=%10.5f" %myval

myval=   1.23457


# Approximation 

Computing in mathematics is first and foremost about _approximation_.  In this class, this word always means

> **Approximation**: Solving one hard problem by turning it into a sequence of simpler problems.  

An example of this idea is encapsulated by the statement

$$
\lim_{n\rightarrow\infty} \left(1 + \frac{1}{n}\right)^{n} = e.
$$

Now $e$ is an irrational number which to an absurd number of digits given by

$$
e = 2.71828182845904523536028747135266249775724709369995
$$

Thus, finding $e$ exactly is difficult.  However, we have a limit statement which says that the numbers $a_{n}$ where

$$
a_{n} = \left(1 + \frac{1}{n}\right)^{n}
$$

get _close to_, or _approximate_, $e$ for very, very large values of $n$.  And as you can see, you can readily use a calculator to determine $a_{n}$ for a given value of $n$.  So while knowing $e$ is difficult, using $a_{n}$ as an approximation allows us to get close.  Let's see this in action by using Python.  

In [21]:
import numpy as np

In [22]:
def exp_approx(n):
    return (1. + 1./n)**(n)

In [23]:
print "e=%1.15f" %exp_approx(1e1) 
print "e=%1.15f" %exp_approx(1e2)
print "e=%1.15f" %exp_approx(1e3)
print "e=%1.15f" %exp_approx(1e4)
print "e=%1.15f" %exp_approx(1e5)
print "e=%1.15f" %exp_approx(1e6)
print "e=%1.15f" %exp_approx(1e7)

e=2.593742460100002
e=2.704813829421528
e=2.716923932235594
e=2.718145926824926
e=2.718268237192297
e=2.718280469095753
e=2.718281694132082


_Problem_: Using the function and printing techniques above, find the 6th digit of $e$.  Said another way, don't just count on the screen.  That isn't really using a computer to its full potential.

In [24]:
# You put in code here 



So we see that, affiliated with an approximation, there is also a notion of _ error _ . Throughout the remainder of this course, by this word we mean 

> **Error**: The difference betweeen a true value or solution and an approximation to it.  

So, if we take as the "true" value of $e$ its 16-digit value

$$
e = 2.718281828459045 
$$

and we use our approximation to $e$ for $n=10^{7}$, then we can find the error between these two things.  

In [25]:
error = np.abs(2.718281828459045 - exp_approx(1e7))
print "The error in floating point notation is: %1.15f" %error
print "The error in scientific notation is: %1.8e" %error

The error in floating point notation is: 0.000000134326963
The error in scientific notation is: 1.34326963e-07


Thus we see that using $n=10^{7}$ gives us 

$$
\left|e - a_{10^{7}} \right| = 1.34326963 \times 10^{-7}
$$

or by using $n=10^{7}$, we get about 6 digits of accuracy after the decimal point.  Note, we have just computed the _absolute error_.  We can also compute the relative error, which in this case would be given by the quantity

$$
\frac{\left|e - a_{10^{7}} \right|}{e}. 
$$

_ Problem _: Compute the relative error of using $a_{10^{7}}$ as an approximation to $e$.  Do this both for floating point and scientific notations.  

In [26]:
# Your turn.  Type your answer in here. 


Unfortunately, not all approximations are created equally.  Said another way, sometimes things do not work as well as we would like.  We see that if we keep trying to use our current approximation for larger values of $n$ we get 

In [27]:
values = [1e7, 1e8, 1e9, 1e10, 1e11, 1e12, 1e13, 1e14] # Build a list of different values of n.

for value in values: # Iterate over the list nvals value by value.
    print "e=%1.15f" %exp_approx(value) # Find the approxmation to e using nval for n.

e=2.718281694132082
e=2.718281798347358
e=2.718282052011560
e=2.718282053234788
e=2.718282053357110
e=2.718523496037238
e=2.716110034086901
e=2.716110034087023


So, let's think about what we just did here.  First, I just introduced a `for` loop on you like whoa!  So what is this doing?  The easiest way to think about it is to realize that all it is doing is sparing us having to cut and paste more since it is exactly equivalent to having used the code

`
print "e=%1.15f" %exp_approx(1e7)
print "e=%1.15f" %exp_approx(1e8)
print "e=%1.15f" %exp_approx(1e9)
print "e=%1.15f" %exp_approx(1e10)
print "e=%1.15f" %exp_approx(1e11)
print "e=%1.15f" %exp_approx(1e12)
print "e=%1.15f" %exp_approx(1e13)
print "e=%1.15f" %exp_approx(1e14)
`

So instead of writing this out over and over again as we did above, we create what is called a _ list _ in Python, containing all the values of n we want to use.  This explains the line of code

`values = [1e7, 1e8, 1e9, 1e10, 1e11, 1e12, 1e13, 1e14]`

Then, the `for` statement iterates over the values within the list by assigning the variable `value` to be a value within `values`, executing the `print` statement.  After this is done, `value` is assigned to the next value of `values`, the `print` statement is executed, and so forth.  This is done until one reaches the end of the list `values`.

So, is the limit statement wrong?  No, it is not.  What we are seeing is the limitations of trying to do math on a computer where every number has a finite amount of precision.  We will talk about this issue in far greater detail later in the course.  Suffice to say though, if we want more digits of $e$, we need a better approximation scheme.  And we get one via our very dear friend, the Taylor Series representation of $e^{x}$, which is given by

$$
e^{x} = \sum_{j=0}^{\infty} \frac{x^{j}}{j!} = 1 + x + \frac{x^{2}}{2!} + \frac{x^{3}}{3!} + \cdots
$$

Note, by $j!$ we mean

$$
j! = j(j-1)(j-2)\cdots(2)(1)(0!), ~ 0! = 1.
$$

So, in essence then, we have a new way to approximate $e$.  We get this by using the Taylor series so that 

$$
e = \sum_{j=0}^{\infty}\frac{1}{j!} = \lim_{n\rightarrow\infty}T_{n}, ~~~~ T_{n}=\sum_{j=0}^{n}\frac{1}{j!}.
$$
So what this says is that we can approxime $e$ using the truncated Taylor series $T_{n}$.  But to do this, we have to take a sum of an arbitrary number of terms in which we have to keep computing ever longer products at every term.  Clearly, a computer should come in handy here.  

The first thing we need to see is that $j!$ is what is called _ recursive _.  What is meant by that is that in order to compute $j!$, we have to compute $(j-1)!$ and then modify this result.  This in words summarizes the identity

$$
j! = j(j-1)!
$$

_ Problem _: Suppose we define the terms $a_{j} = \frac{1}{j!}$.  Show, for $j\geq 0$ that 

$$
a_{j} = \frac{a_{j-1}}{j}.
$$

Now, why fuss about this?  Well, as we are about to see, how we write the math and how we write the code can look very different.  They are always related, but frankly, they are different languages, and so in effect we are obliged to translate.  Let me show you what I mean.  Below is the code I would use to compute the partial sums above.

In [28]:
def exp_sum(tol): # A user inputs a specified tolerance given by the variable tol.
    Tj = 1. # We initialize the sum.
    aj = 1. # We initialize the term.
    j = 1 # We initialize the term number we are computing.  
    while np.abs(aj)>=tol: # Keep updating and adding terms until a term is smaller than the tolerance tol.
        aj *= 1./j # Updates the terms of the sum.
        Tj += aj # Adds the next term to the sum.
        j += 1 # Updates which term we are computing.
    return Tj

We will dissect this code in a couple of lectures.  But we can see what results we get from it.  And, yup, you guessed it, I'm going to give you more code to wrap your head around.  

In [29]:
tolvals = [1e-1, 1e-2, 1e-3, 1e-4, 1e-5, 1e-6, 1e-7] # Decide on the tolerance values we want to test.

for tolval in tolvals: # Iterate over the values within the list tolvals value by value
    print "e=%1.15f" %exp_sum(tolval) # Find an approximation to e for the given tolerance value tol

e=2.708333333333333
e=2.716666666666666
e=2.718253968253968
e=2.718278769841270
e=2.718281525573192
e=2.718281801146385
e=2.718281826198493


Let's see what the corresponding error looks like.  

In [30]:
etrue = 2.718281828459045
tolvals = [1e-1, 1e-2, 1e-3, 1e-4, 1e-5, 1e-6, 1e-7,1e-8,1e-9,1e-10,1e-11] # Decide on the tolerance values we want to test.

for tolval in tolvals:
    print "The absolute error using a tolerance of %1.0e is: %1.5e" %(tolval,np.abs(etrue - exp_sum(tolval))) 

The absolute error using a tolerance of 1e-01 is: 9.94850e-03
The absolute error using a tolerance of 1e-02 is: 1.61516e-03
The absolute error using a tolerance of 1e-03 is: 2.78602e-05
The absolute error using a tolerance of 1e-04 is: 3.05862e-06
The absolute error using a tolerance of 1e-05 is: 3.02886e-07
The absolute error using a tolerance of 1e-06 is: 2.73127e-08
The absolute error using a tolerance of 1e-07 is: 2.26055e-09
The absolute error using a tolerance of 1e-08 is: 1.72876e-10
The absolute error using a tolerance of 1e-09 is: 1.22857e-11
The absolute error using a tolerance of 1e-10 is: 8.14904e-13
The absolute error using a tolerance of 1e-11 is: 5.01821e-14


So as we can see, we have a vastly superior method for computing $e$ by using the truncated Taylor series approach.  With that being said, let's now spend some time dwelling on Taylor Series in greater detail.  